server/data_common/driver.py

```diff
@@ -0,0 +1,319 @@
+        try:
+            shape = self.get_X_array_shape()
+            obs_mask_A = self._axis_filter_to_mask(Axis.OBS, obsFilterA["obs"], shape[0])
+            obs_mask_B = self._axis_filter_to_mask(Axis.OBS, obsFilterB["obs"], shape[0])
+        except (KeyError, IndexError) as e:
+            raise FilterError(f"Error parsing filter: {e}") from e
+        if top_n is None:
+            top_n = DEFAULT_TOP_N
+
+        result = diffexp_ttest(self, obs_mask_A, obs_mask_B, top_n, self.config.diffexp_lfc_cutoff)
+
+        try:
+            return jsonify_numpy(result)
+        except ValueError:
+            raise JSONEncodingValueError("Error encoding differential expression to JSON")
+
+    def layout_to_fbs_matrix(self):
+        """ same as layout, except returns a flatbuffer """
+        """
+        return all embeddings as a flatbuffer, using the cellxgene matrix fbs encoding.
+
+        * returns only first two dimensions, with name {ename}_0 and {ename}_1,
+          where {ename} is the embedding name.
+        * client assumes each will be individually centered & scaled (isotropically)
+          to a [0, 1] range.
+        * does not support filtering
+
+        """
+
+        embeddings = self.get_embedding_names()
+        layout_data = []
+        with ServerTiming.time(f'layout.query'):
+            for ename in embeddings:
+                embedding = self.get_embedding_array(ename, (slice(None), slice(0, 2)))
+
+                # scale isotropically
+                min = embedding.min(axis=0)
+                max = embedding.max(axis=0)
+                scale = np.amax(max - min)
+                normalized_layout = (embedding - min) / scale
+
+                # translate to center on both axis
+                translate = 0.5 - ((max - min) / scale / 2)
+                normalized_layout = normalized_layout + translate
+
+                normalized_layout = normalized_layout.astype(dtype=np.float32)
+                layout_data.append(pd.DataFrame(normalized_layout, columns=[f"{ename}_0", f"{ename}_1"]))
+
+        with ServerTiming.time(f'layout.encode'):
+            if layout_data:
+                df = pd.concat(layout_data, axis=1, copy=False)
+            else:
+                df = pd.DataFrame()
+            fbs = encode_matrix_fbs(df, col_idx=df.columns, row_idx=None)
+
+        return fbs
+
+    @abstractmethod
+    def get_name(self):
+        """return a string name for this data engine"""
+        pass
+
+    @abstractmethod
+    def get_library_versions(self):
+        """return a dictionary of library name to library versions"""
+        pass
+
+    @abstractmethod
+    def get_embedding_names(self):
+        """return a list of embedding names"""
+        pass
+
+    @abstractmethod
+    def get_embedding_array(self, ename, items=None):
+        """return an numpy array for the given embedding name"""
+        pass
+
+    @abstractmethod
+    def get_X_array(self, items=None):
```

Comment:
this (and other) API that take indices need to declare what indexing format they support and if it is inclusive or exclusive.

I believe that you are currently only accepting an ndarray or list of scalar indices (eg, [0, 8, 33]), one for each dimension.

So this function should also most likely accept two args:   obs_items, var_items

Reply:
Fixed.  the methods in the derived classes take in 2 args, so this was inconsistent.  Also added comments.